internal/pipeline/pipeline_function.go

```diff
@@ -0,0 +1,205 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package pipeline provides struct definitions for Pipeline and utility
+// methods to read and write a pipeline resource.
+package pipeline
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/pipeline/runtime"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// Function defines an item in the pipeline function list
+type Function struct {
+	// `Image` is the path of the function container image
+	// Image name can be a "built-in" function: kpt can be configured to use a image
+	// registry host-path that will be used to resolve the full image path in case
+	// the image path is missing (Defaults to gcr.io/kpt-functions-trusted).
+	// For example, the following resolves to gcr.io/kpt-functions-trusted/patch-strategic-merge.
+	//		image: patch-strategic-merge
+	Image string `yaml:"image,omitempty"`
+
+	// `Config` specifies an inline k8s resource used as the function config.
+	// Config, ConfigPath, and ConfigMap fields are mutually exclusive.
+	Config yaml.Node `yaml:"config,omitempty"`
+
+	// `ConfigPath` specifies a relative path to a file in the current directory
+	// containing a K8S resource used as the function config. This resource is
+	// excluded when resolving 'sources', and as a result cannot be operated on
+	// by the pipeline.
+	ConfigPath string `yaml:"configPath,omitempty"`
+
+	// `ConfigMap` is a convenient way to specify a function config of kind ConfigMap.
+	ConfigMap map[string]string `yaml:"configMap,omitempty"`
+}
+
+// Validate will validate all fields in function.
+func (f *Function) Validate() error {
+	err := validateFunctionName(f.Image)
+	if err != nil {
+		return fmt.Errorf("'image' is invalid: %w", err)
+	}
+
+	var configFields []string
+	if f.ConfigPath != "" {
+		if err := validatePath(f.ConfigPath); err != nil {
+			return fmt.Errorf("'configPath' %q is invalid: %w", f.ConfigPath, err)
+		}
+		configFields = append(configFields, "configPath")
+	}
+	if len(f.ConfigMap) != 0 {
+		configFields = append(configFields, "configMap")
+	}
+	if !isNodeZero(&f.Config) {
+		configFields = append(configFields, "config")
+	}
+	if len(configFields) > 1 {
+		return fmt.Errorf("following fields are mutually exclusive: 'config', 'configMap', 'configPath'. Got %q",
+			strings.Join(configFields, ", "))
+	}
+
+	return nil
+}
+
+// runner returns a fnRunner from the image and configs of
+// this function.
+func (f *Function) runner() (kio.Filter, error) {
+	config, err := f.config()
+	if err != nil {
+		return nil, err
+	}
+	return &fnRunner{
+		fn: &runtime.ContainerFn{
+			Image: f.Image,
+		},
+		fnConfig: config,
+	}, nil
+}
+
+func (f *Function) config() (*yaml.RNode, error) {
+	var node *yaml.RNode
+	switch {
+	case f.ConfigPath != "":
+		file, err := os.Open(f.ConfigPath)
```

Comment:
ConfigPath is a slash-delimited relative path so you can't just os.Open() on it.